### DISCUSSION TITLE
[Proposal] Repository structure for framework tests

### DISCUSSION THREAD

**Discussion by SamMousa** [Codeception/Codeception#6207] [Ideas]
#### What are you trying to achieve?
Codeception contains several framework modules. Ideally these framework modules should have separate test suites so that tests for discovered issues in a framework module can be created.
This will allow us to prevent regressions and increase reliability of the modules.

#### Proposal
I propose a common repository structure for framework module tests:
1. Each framework creates one test repository, this could be under the Codeception organization but this is not a technical requirement.
2. During framework module testing, the test script will checkout the test repository including git submodules.
3. Composer install is executed for the test repo and its submodules.
4. Codeception tests are ran.

A test repo needs to have a `codeception.yml` in its root and this must include all `codeception.yml` files from submodules. (Otherwise they won't be executed).

### Directory structure:
This is an example from the [yii2-tests](https://github.com/Codeception/yii2-tests) repo.

```bash
.
├── bootstrap.php
├── cases
│   ├── simple
│   └── sqlite
├── codeception.yml
├── composer.json
├── .gitignore
├── .gitmodules
├── readme.md
└── .travis.yml
```

The requirement on the repo is that all tests are included in the toplevel codeception file; in the case of Yii2 tests there are several test cases that use different application configs, so their tests are included in the main file:
```yml
paths:
  tests: .
  log: _output
  data: _data
  support: _support
modules:
  - Asserts
settings:
  bootstrap: bootstrap.php
  colors: true
  memory_limit: 1024M
include:
  - cases/*
  - submodules/*
```
Note that tests from submodules are included as well.
For submodules the travis CI script will take care of installing composer dependencies via
`git submodule foreach composer install --prefer-dist --no-dev`

### Conclusion
I think some guidelines that are documented would be great to further improve module quality as well as lower the threshold for new framework modules to get added.
This is just the first draft of my proposal and any feedback is greatly appreciated.
For example, if we don't really need submodule support we are probably better off without it as it will probably increase test times considerably.

**Comment by DavertMik** [comment]
Thanks, that's a good idea and a good start. 
I like the overall idea of restructuring tests. And I like the idea of making the separate repository but I'm not sure about the time and efforts we will put into this restructuing. 

But I think the overall idea should be moving out framework modules into separate repos so the module and tests would be in the same repo. And we could provide bugfix with a test with one pull request

**Comment by SamMousa** [comment]
> I like the overall idea of restructuring tests. And I like the idea of making the separate repository but I'm not sure about the time and efforts we will put into this restructuring.

No need to prioritize this, just something to have on the todo list.


> But I think the overall idea should be moving out framework modules into separate repos so the module and tests would be in the same repo. And we could provide bugfix with a test with one pull request

I definitely think that's the way to go!

**Comment by eXorus** [comment]
I like the idea also, thanks @SamMousa
